lib/charms/vault_k8s/v0/vault_autounseal.py

```diff
@@ -0,0 +1,518 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Library for the vault-autounseal relation.
+
+This library contains the Requires and Provides classes for handling the
+vault-autounseal interface.
+
+The provider side of the interface is responsible for enabling the vault
+transit engine and creating the necessary keys and policies for an external
+vault to be able to autounseal itself.
+
+The requirer side of the interface is responsible for retrieving the necessary
+details to autounseal the vault instance, and configuring the vault instance to
+use them.
+
+## Getting Started
+
+From a charm directory, fetch the library using `charmcraft`:
+
+```shell
+charmcraft fetch-lib charms.vault_k8s.v0.vault_autounseal
+```
+
+### Provider charm
+
+The provider charm is the charm that provides a Vault instance that can be
+used to autounseal other Vault instances via the Vault transit backend.
+
+Add the following to `metadata.yaml`:
+
+```yaml
+provides:
+  vault-autounseal-provides:
+    interface: vault-autounseal
+```
+
+### Requirer charm
+
+The requirer charm is the charm that wishes to autounseal a Vault instance via
+the Vault transit backend.
+
+Add the following to `metadata.yaml`:
+
+```yaml
+requires:
+  vault-autounseal-requires:
+    interface: vault-autounseal
+    limit: 1
+```
+
+### Integration
+
+You can integrate both charms by running:
+
+```bash
+juju integrate <vault a>:vault-autounseal-provides <vault b>:vault-autounseal-requires
+```
+
+where `vault a` is the Vault app which will provide the autounseal service, and
+`vault b` is the Vault app which will be configured for autounseal via `vault a`.
+"""
+
+import logging
+from dataclasses import dataclass
+from typing import Any, Dict, List
+
+import ops
+from interface_tester import DataBagSchema
+from ops import Relation, RelationDataContent, SecretNotFoundError, model  # type: ignore
+from pydantic import BaseModel, Field, ValidationError
+
+# The unique Charmhub library identifier, never change it
+LIBID = "c33e0a12506444e2b644ac2893ac9394"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 3
+
+
+class LogAdapter(logging.LoggerAdapter):
+    """Adapter for the logger to prepend a prefix to all log lines."""
```

Comment:
Have you considered documenting the attrs?

Reply:
Do you have any suggestion?
I don't have anything in mind that would add value here.

Reply:
We document the classs attributes as part of the docstring, so I was wondering if you are following the same convention. You can see an example [here](https://github.com/canonical/synapse-operator/blob/main/src/charm_state.py#L128)

Reply:
In general, we do, but if the documentation doesn't add much value, we don't enforce it. In this case, I don't see much value in documenting these arguments.